client/asset/dcr/dcr.go

```diff
@@ -1654,11 +1899,36 @@ func (dcr *ExchangeWallet) fund(enough func(sum uint64, size uint32, unspent *co
 		return nil, nil, 0, 0, err
 	}
 
-	sum, sz, coins, spents, redeemScripts, err := tryFund(utxos, enough)
+	avail := sumUTXOs(utxos)
+	if keep > avail { // skip utxo selection if we can't possibly make reserves
+		return nil, nil, 0, 0, asset.ErrInsufficientBalance
+	}
+
+	sum, extra, sz, coins, spents, redeemScripts, err := tryFund(utxos, enough)
 	if err != nil {
 		return nil, nil, 0, 0, err
 	}
 
+	if avail-sum+extra < keep {
+		if extra != 0 { // there was usable change from enough, just not enough
+			return nil, nil, 0, 0, asset.ErrInsufficientBalance
+		} // else caller probably not granting change
+		// NOTE: tryFund above may not have tried 0-conf, but we may now.
+
+		dcr.log.Debugf("Trying again with some UTXOs excluded for required reserves of %v DCR",
+			toDCR(keep))
+
+		// Try again with some utxos taken out of the mix for keep. Select these
+		// with the objective of being as close to the amount as possible,
+		// unlike tryFund that minimizes the number of UTXOs chosen.
+		kept := leastOverFund(keep, utxos)
+		utxos = utxoSetDiff(utxos, kept)
+		sum, _, sz, coins, spents, redeemScripts, err = tryFund(utxos, enough)
```

Comment:
If we get here, it is because the `enough` func is very likely not granting any spendable change i.e. no split tx.  Note the `if extra != 0 {` check above.  The logic in this commit is to try again but fist pick out utxos in the amount of at least `keep` before even calling `tryFund` -- that guarantees `keep` will be met but it can then fail if the pruned utxo set is insufficient for the order's `enough` func.

However (and I apologize for the moving target), after we chatted on matrix this morning, I switched the order of these attempts so that it _first_ sets aside utxos (using just the pruned set with `tryFund`), and if that fails to leave enough for to fund, it will then try with the full set and check if `extra` allowed it to satisfy `keep`: https://github.com/decred/dcrdex/pull/2103/commits/9f6ad97a87d0a29a7f9166efec3b37efa19cddc1